### ISSUE TITLE
add vector to diagonal of matrix

### ISSUE THREAD

**Issue by bob-carpenter** [stan-dev/math#411]
#### Summary:

Add function with signature

```
matrix add_to_diag(vector v, matrix u)
```

defined by

```
add_to_diag(v, u) = diag_matrix(v) + u
```

I think it's better to just have a single function than to also try to support `diag_add(matrix, vector)`.  

It'd also be useful to have a scalar version that adds a scalar to the diagonal of a matrix,

```
add_to_diag(real c, matrix u);
```

with the behavior being to add `c` to the diagonal elements of `u`, which is inefficient (and clunky) even with the above,

```
add_to_diag(c, u) = add_to_diag(rep_vector(c, rows(u)), u)
```
#### Description:

A specialized function avoids the inefficiencies of the current forms (lots of zeroes).

For the vector case, a simple template is probably sufficient as it's just a bunch of adds.

For the scalar case, a custom autodiff would help because the scalar's reused for each diagonal entry and so the partial is the sum of the diagonals.
#### Current Version:

v2.12.0

**Comment by alinarezrangel** [comment]
Hi @bob-carpenter 

I would love to contribute to the project in this task, I will begin to write the function and send my request in a time.

**Comment by bob-carpenter** [comment]
That'd be great.  Let us know if you need help.  Did you find
the how to write new functions for Stan doc on our stan-dev/stan wiki?

**Comment by alinarezrangel** [comment]
Yes, thanks.

**Comment by bob-carpenter** [comment]
I realize there's an issue with this operation.

If we write

```
matrix diag_add(vector u, matrix v);
```

there's no way to make it less than an O(M x N) operation where
M x N is the size of the matrix.  Whereas writing this:

```
for (k in 1:K)
  v[k, k] = v[k, k] + u[k];
```

it is only an O(K) operation.  

Copies aren't very expensive, but still, I hate to see this kind
of blow-up.

We don't have anything other than pass-by-value (really constant
reference at the C++ level), so there's no way to do something like
the void return that modifies v.

What we really want is good template expressions on the C++ side
and assignments like this in Stan:

```
v += diag_matrix(u);
```

But that's a long way off in terms of our matrix coding.

**Comment by alinarezrangel** [comment]
I think that the function `add_to_diag` (or `diag_add`) can return a vector, that will be transformed into a diagonal matrix when it's required. By this, we can save the `O(K)` performance but without using a copy loop:

```
# We can define diag_matrix as an alias for vector

# Is a simple vector add:
diag_matrix diag_add(vector u, diag_matrix v) = v + u

matrix diag_to_mat(diag_matrix v) = diag_matrix(v)

# For example:

diag_matrix my_diagonal_matrix;
vector my_vector = [1, 3, 2, 10, 6, 4 , 7]

# We can use:
diag_matrix my_dmatrix = diag_add(my_vector, my_diagonal_matrix)
matrix my_matrix = diag_to_mat(my_dmatrix)

# Use my_matrix
```

Using this, we can optimize all diagonal operations to a vector, and we can add custom function for transform a `diag_matrix` to a, for example, anti-diagonal matrix instead:

```
# Something like this:
diag_matrix my_matrix = (some operations...)

matrix my_result = diag_to_anti_matrix(my_matrix)
```

**Comment by bob-carpenter** [comment]
Loops aren't the problem, nor is aliasing going to help.

The fundamental problem is that if we write a function
with this signature in Stan:

  matrix add_to_diag(vector, matrix);

and define the return to be equivalent to 

  diag_matrix(vector) + matrix

there's no way to get around copying the off-diagonal entries
in the input matrix to the output.

We don't need a new function to add two vectors---we
already have that.  And we have diag_matrix that will convert
a vector to a diagonal matrix.  Now we could code the above
so it's more efficient than diag_matrix(vector) + matrix,
because we don't actually have to build the diag_matrix zero
entries, but we still have to copy.

I think what you may be thinking about is something like
expression templates, which we don't support yet.  See:

https://en.wikipedia.org/wiki/Expression_templates

Then we could use an expression template for add_to_diag.
Eigen supports a lot of this already, but we're using plain
base matrices in Stan.  It'd be great to upgrade all of our
code to support expression templates, but it'd be a ton of
work.

Right now, it'd be useful to just write the above function
as so:

template <typename T1, typename T2>
Matrix<typename promote_args<T1, T2>::type, -1, -1>
add_diag(const Matrix<T1, -1, -1>& a,
         const Matrix<T2, 1, -1> & v) {
  check_square(...)
  check_matching_sizes(...)
  Matrix<typename promote_args<T1, T2>::type, -1, -1> a_plus_v(a);
  for (int k = 0; k < v.rows(); ++k)
    a_plus_v(k, k) += v(k);
  return a_plus_v;
}

We can't quite get away with using Eigen's expression templates
directly because of the potentially mixed type of T1 and T2.

All Stan functions need to work for all scalar-type instantiations
so as not to break the language.
 
- Bob

> On Oct 25, 2016, at 4:07 PM, Alejandro Linarez Rangel <notifications@github.com> wrote:
> 
> I think that the function add_to_diag (or diag_add) can return a vector, that will be transformed into a diagonal matrix when it's required. By this, we can save the O(K) performance but without using a copy loop:



> # We can define diag_matrix as an alias for vector

> # Is a simple vector add:
> diag_matrix diag_add(vector u, diag_matrix v) = v + u
> 
> matrix diag_to_mat(diag_matrix v) = diag_matrix(v)
> 
> # For example:
> 
> diag_matrix my_diagonal_matrix;
> vector my_vector = [1, 3, 2, 10, 6, 4 , 7]
> 
> # We can use:
> diag_matrix my_dmatrix = diag_add(my_vector, my_diagonal_matrix)
> matrix my_matrix = diag_to_mat(my_dmatrix)
> 
> # Use my_matrix
> 
> Using this, we can optimize all diagonal operations to a vector, and we can add custom function for transform a diag_matrix to a, for example, anti-diagonal matrix instead:
> 
> # Something like this:
> diag_matrix my_matrix = (some operations...)
> 
> matrix my_result = diag_to_anti_matrix(my_matrix)

- Bob

**Comment by alinarezrangel** [comment]
In that case, we can start to work on these characteristics of the stan language and then, make the functions `diag_*`. I can help on change the language for support expression templates and other required characteristics.

**Comment by andrjohns** [comment]
Closed via #871
